ouroboros-network/src/Ouroboros/Network/PeerSelection/LedgerPeers.hs

```diff
@@ -0,0 +1,140 @@
+    LedgerPeersConsensusInterface (..),
+    RelayAddress (..),
+    PoolStake (..),
+    AccPoolStake (..),
+    TraceLedgerPeers (..),
+    pickPeers,
+    accPoolStake,
+
+    Socket.PortNumber
+    ) where
+
+
+import           Control.Monad.Class.MonadSTM
+import           Control.Tracer (Tracer, traceWith)
+import qualified Data.IP as IP
+import           Data.List (foldl')
+import           Data.List.NonEmpty (NonEmpty (..))
+import qualified Data.List.NonEmpty as NonEmpty
+import qualified Data.Map.Strict as Map
+import           Data.Map.Strict (Map)
+import           Data.Ratio
+import           Data.Word
+import qualified Network.Socket as Socket
+import           System.Random
+
+import           Cardano.Slotting.Slot (SlotNo)
+import           Ouroboros.Network.PeerSelection.RootPeersDNS (DomainAddress (..))
+
+import           Text.Printf
+
+newtype LedgerPeersConsensusInterface m = LedgerPeersConsensusInterface {
+      lpGetPeers :: SlotNo -> STM m (Maybe [(PoolStake, NonEmpty RelayAddress)])
+    }
+
+-- | Trace LedgerPeers events.
+data TraceLedgerPeers =
+      PickedPeer !RelayAddress !AccPoolStake ! PoolStake
+      -- ^ Trace for a peer picked with accumulated and relative stake of its pool.
+    | PickedPeers !Word16 ![RelayAddress]
+      -- ^ Trace for the number of peers we wanted to pick and the list of peers picked.
+    | FetchingNewLedgerState !Int
+      -- ^ Trace for fetching a new list of peers from the ledger. Int is the number of peers
+      -- returned.
+
+
+instance Show TraceLedgerPeers where
+    show (PickedPeer addr ackStake stake) =
+        printf "PickedPeer %s ack stake %s ( %.04f) relative stake %s ( %.04f )"
+            (show addr)
+            (show $ unAccPoolStake ackStake)
+            (fromRational (unAccPoolStake ackStake) :: Double)
+            (show $ unPoolStake stake)
+            (fromRational (unPoolStake stake) :: Double)
+    show (PickedPeers n peers) =
+        printf "PickedPeers %d %s" n (show peers)
+    show (FetchingNewLedgerState cnt) =
+        printf "Fetching new ledgerstate, %d registered pools"
+            cnt
+
+
+data RelayAddress = RelayAddressDomain DomainAddress
+                  | RelayAddressAddr IP.IP Socket.PortNumber
+                  deriving (Show, Eq, Ord)
+
+-- | The relative stake of a stakepool in relation to the total amount staked.
+-- A value in the [0, 1] range.
+--
+newtype PoolStake = PoolStake { unPoolStake :: Rational }
+  deriving (Eq, Fractional, Num, Ord, Show)
+
+-- | The accumulated relative stake of a stake pool, like PoolStake but it also includes the
+-- relative stake of all preceding pools. A value in the range [0, 1].
+--
+newtype AccPoolStake = AccPoolStake { unAccPoolStake :: Rational }
+    deriving (Eq, Num, Ord)
+
+-- | Convert a list of pools with stake to a Map keyed on the accumulated stake.
+--
+accPoolStake :: [(PoolStake, NonEmpty RelayAddress)]
+             -> Map AccPoolStake (PoolStake, NonEmpty RelayAddress)
```

Comment:
I don't understand the significance of the order in `[(PoolStake, NonEmpty RelayAddress)]`.  The more to the right of the list is a pool the less`AccPoolStake` it will have.  What's the assumption / semantics of the list?

Reply:
Consensus provides an ordered list, but it has no effect on the picking scheme.

Reply:
I think I got it now; it would be good if there was a paragraph how the picking scheme works, certainly that would help me read the code.

Reply:
Agreed. will add something.

Reply:
Ok, let me know when you'll be done; Apart from this it's _LGTM_.